lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -965,6 +965,85 @@ class MongoClientInterface {
         ], cb);
     }
 
+    /**
+     * gets object metadata for a list of objects
+     * @param {String} bucketName bucket name
+     * @param {Array} objects array of objects
+     * @param {Object} log logger
+     * @param {Function} callback callback
+     * @return {undefined}
+     */
+    getObjects(bucketName, objects, log, callback) {
+        const c = this.getCollection(bucketName);
+        let vFormat = null;
+        if (!Array.isArray(objects)) {
+            return callback(errors.InternalError.customizeDescription('objects must be an array'));
+        }
+        // Function to process each document
+        const processDoc = (doc, objName, params, key, cb) => {
+            const versionIdValue = params && params.versionId ? params.versionId : undefined;
+            if (!doc && versionIdValue) {
+                // If no document and a version ID is provided, return an error.
+                return cb(null, {
+                    err: errors.NoSuchKey,
+                    doc: null,
+                    versionId: versionIdValue,
+                    key,
+                });
+            }
+            // If no master found then object is either non existent or last
+            // version is delete marker
+            if (!doc || doc.value.isPHD) {
+                return this.getLatestVersion(c, objName, vFormat, log, (err, _doc) => cb(null, {
+                    err,
+                    doc: _doc || null,
+                    versionId: versionIdValue,
+                    key,
+                }));
+            }
+            MongoUtils.unserialize(doc.value);
+            return cb(null, {
+                err: null,
+                doc: doc.value,
+                versionId: versionIdValue,
+                key,
+            });
+        };
+        return this.getBucketVFormat(bucketName, log, (err, _vFormat) => {
+            if (err) {
+                return callback(err);
+            }
+            vFormat = _vFormat;
+            // Create keys, maintaining the context with each key
+            const keysAndObjects = objects.map(({ key: objName, params }) => {
+                const _key = params && params.versionId
+                    ? formatVersionKey(objName, params.versionId, vFormat)
+                    : formatMasterKey(objName, vFormat);
+                return { key: _key, objName, params };
+            });
+            // Extract keys and find documents
+            const keys = keysAndObjects.map(o => o.key);
+            return c.find({
+                _id: { $in: keys },
```

Comment:
mongo query size is limited by the max BSON object length (16MB) : so we need to make sure we don't hit that limit...

Now multi-delete is limited to 1000 objects at a time, and object name is limited to 1024 characters : so we should be fine, worse-case at around 1.1MB.

We may want to check the number of objects passed to `getObjects`, though, to be on the safe side.

Reply:
Done in https://github.com/scality/Arsenal/pull/2134/commits/cc75b89115a7a8d4f26417fde5103f80421bedef#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR979